### ISSUE TITLE
Interpretation callback - best and worst indices of a loss for loader

### ISSUE THREAD

**Issue by hristo-vrigazov** [catalyst-team/catalyst#950]
## 🚀 Feature Request
Hi, I think we already discussed this in Slack, but the basic idea is the following: there is a callback called e.g `LossInterpretationCallback` which shows the best/worst examples for a loss in the Tensorboard.

### Motivation
Motivation: quick error analysis, to gain a quick intuition on when is the model doing great and when it is doing badly. Since it is just a callback, users can always just not include it, there is no risk for other features.

### Proposal
My proposal is to create a class called `LossInterpretationCallback` located either under `contrib` or under `dl.callbacks`, that for every loader that uses a `SequentialSampler` (otherwise we cannot know the original index in the dataset) computes a given criterion reduced per sample, and then sorts it, and plot top k / bottom k examples to the Tensorboard, and also serializes the interpretations for example in a `loss_interpretation` directory, so that they can be further inspected later.

### Alternatives
Several similar, but slightly different callbacks exist:

* [ShowPolarBatchesCallback](https://github.com/BloodAxe/pytorch-toolbelt/blob/c51c41f12503d76f6f3bd452aa95a7765628db07/pytorch_toolbelt/utils/catalyst/visualization.py#L34) in `pytorch-toolbelt` - shows only the best one and worst one *batches*, not samples. Slightly different, still very useful
* [InterpretationCallback](https://dnncool.readthedocs.io/en/latest/dnn_cool.html?highlight=InterpretationCallback#dnn_cool.catalyst_utils.InterpretationCallback) in `dnn.cool` -  the same thing, but it forces you to use multi-task learning classes, is awkward if you have just one output.

### Additional context
Here's for example how would the minimal classification example change:

```python
#%%

import torch
from torch.utils.data import DataLoader, TensorDataset
from catalyst import dl

#%%
# sample data
num_samples, num_features, num_classes = int(1e4), int(1e1), 4
X = torch.rand(num_samples, num_features)
y = (torch.rand(num_samples, ) * num_classes).to(torch.int64)

# pytorch loaders
dataset = TensorDataset(X, y)
loader = DataLoader(dataset, batch_size=32, num_workers=1)
loaders = {"train": loader, "valid": loader}

# model, criterion, optimizer, scheduler
model = torch.nn.Linear(num_features, num_classes)
criterion = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters())
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, [2])

#%%
callbacks = [dl.AccuracyCallback(num_classes=num_classes),
             dl.LossInterpretationCallback(torch.nn.CrossEntropyLoss(reduction='none'))]
#%%
# model training
runner = dl.SupervisedRunner()
runner.train(
    model=model,
    criterion=criterion,
    optimizer=optimizer,
    scheduler=scheduler,
    loaders=loaders,
    logdir="./logdir",
    num_epochs=3,
    callbacks=callbacks
)
```

**Comment by hristo-vrigazov** [comment]
I forgot to mention that I am willing to contribute it

**Comment by Scitator** [comment]
so, summon @BloodAxe as `pytorch-toolbelt` author :)
I think it could be also interesting for you.

**Comment by Scitator** [comment]
Hi @hristo-vrigazov, nice to see your proposal :)
I like the idea, but what I am not sure about is... `SequentialSampler` - it's quite uncommon case to use sequential sampling for you data...
what do you think, about serialisation/saving the whole input dict from `Dataset`? In this case you could do some thing like
```python

class MyDataset(Dataset):
  def __get__():
    return {"features": features, "targets": targets, "some-meta": {}, "index": index}
```
and you could find your original sample with the help of your extra data - ``"some-meta": {}, "index": index``.
This solution requires you to add extra keys to our Dataset, but could work with any Sampler/Loader

**Comment by hristo-vrigazov** [comment]
Is it that uncommon? I very often do not shuffle my validation and test set, so they end up using `SequentialSampler` ( https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L263 ), and usually, I am interested in interpreting their results. Yes, including the index in the dataset would work for every sampler, but then it requires additional work for every user (and also forces them to use dict-style output), so this is the minus of this approach. Maybe we can make this configurable, e.g have a parameter called `index_key` that is optional and if it is not supplied, then it works only for `SequentialSampler`-based loaders. If it is supplied, then we obviously get it and store it. What do you think?

**Comment by Scitator** [comment]
🤔 Speaking about validation... you are right, you don't need any shuffle there -> `SequentialSampler` -> index solution would work!

**Comment by Scitator** [comment]
In this case looks like we really could implement such useful callback - it would be great,
btw, @bagxi @Ditwoo could you please share your thoughts on this feature and check our proposals?
